{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "e74c7385",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''\n",
    "GLOBAL ALIGNMENT WITH AFFINE GAPS\n",
    "'''\n",
    "\n",
    "def read_single_fasta(filename):\n",
    "    \"\"\"\n",
    "    Reads in a single fasta file (only contains one sequence) and returns\n",
    "    the sequence from the file.\n",
    "    Input: filename (String) - the name of the fasta file to read. The\n",
    "    sequence may be on multiple lines in the fasta file.\n",
    "    Output: sequence (String) - a single string representing the sequence in\n",
    "    the fasta file.\n",
    "    Code from HW1 solutions\n",
    "    \"\"\"\n",
    "    f = open(filename, 'r')\n",
    "    sequence = \"\"\n",
    "    for line in f:\n",
    "        if not line.startswith(\">\"):\n",
    "            #only include non headers\n",
    "            sequence += line.rstrip()\n",
    "    f.close()\n",
    "    sequence = sequence.upper()\n",
    "    return sequence\n",
    "'''\n",
    "READING PENALTY FILE\n",
    "'''\n",
    "\n",
    "def print_file(filename):\n",
    "    '''\n",
    "    This function just prints to the screen all contents within a file.\n",
    "    Input: filename (String) - the name of the file to print.\n",
    "    '''\n",
    "    file = open(filename, 'r') # 'r' says we are reading the file\n",
    "\n",
    "    print(\"Printing contents of following file:\",filename)\n",
    "    for line in file: # this is an easy way to iterate through a file\n",
    "        line = line.rstrip() # all lines have newline character at the end.\n",
    "        print(line)\n",
    "    file.close() #always close the file after opening it\n",
    "\n",
    "    print(\"\\n\\n\\n\") #print a few lines of white space\n",
    "\n",
    "\n",
    "def read_file(filename):\n",
    "    '''\n",
    "    This function takes in a filename and returns a tuple\n",
    "    containing the four different scoring parameters.\n",
    "    Input: filename (string) - the name of the file to open.\n",
    "    Return: scores (tuple of ints) - this tuple of length 4\n",
    "            contains the following values in the following order:\n",
    "            match score, mismatch score, gap-extend, gap-open\n",
    "    '''\n",
    "    file = open(filename, 'r') # 'r' says we are reading the file\n",
    "\n",
    "    file.readline() #skip past header line\n",
    "\n",
    "    line = file.readline() #reads next line, returns a string.\n",
    "    line = line.rstrip() #remove all white space, return characters, etc\n",
    "                         #on right hand side of string.\n",
    "\n",
    "    file.close() #done with file, so close it\n",
    "\n",
    "    vals = line.split(\"\\t\") #turn string into list of items originally\n",
    "                            #separated by tabs\n",
    "\n",
    "    scores = tuple(vals) #convert from list (mutable) to tuple (immutable)\n",
    "\n",
    "    return scores\n",
    "\n",
    "def write_file(scores, filename):\n",
    "    '''\n",
    "    This function takes in a tuple of scores and a filename\n",
    "    and saves those scores to file as a new scoring file.\n",
    "    Input: scores (tuple of ints) - a tuple of length 4 containing match,\n",
    "            mismatch, gap-open and gap extend values.\n",
    "           filename (String) - the name of the file to create.\n",
    "    Return: None\n",
    "    '''\n",
    "    file = open (filename, 'w') #w indicates we are writing to a file\n",
    "    header = \"#match\\tmismatch\\tgap-extend\\tgap-open\\n\" #need to add newline at end\n",
    "    file.write(header) #This adds header to the file\n",
    "\n",
    "    #Build string of tab separated scores.\n",
    "    vals = \"\\t\".join(scores) #creates a tab sep string of everything in scores.\n",
    "    vals = vals + \"\\n\" #add newline at end\n",
    "    file.write(vals) #This adds vals to the file\n",
    "    file.close() #You have to close a file in order for everything you wrote\n",
    "                 #into it to show up\n",
    "        \n",
    "#READS FILE TO GET PENALTIES\n",
    "def scoringfile(scfilename):\n",
    "    '''\n",
    "    Read in a scoring file provided as first command line parameter.\n",
    "    Print scores to the screen.  Then return the final values as 1 string seperates by semicolons.\n",
    "    '''\n",
    "    arg = scfilename\n",
    "    \n",
    "    num_args = len(arg) - 1 #retrieve number of command line arguments\n",
    "    if (num_args < 1):\n",
    "        print(\"Program requires a file name.\")\n",
    "        \n",
    "\n",
    "    filename = arg #first argument should be name of scoring file.\n",
    "                           #this assumes file is in same directory as this code.\n",
    "\n",
    "    #print contents of file to screen\n",
    "    print_file(filename)\n",
    "\n",
    "    #print all scores to the screen\n",
    "    scores = read_file(filename)\n",
    "    print(\"Scores contained in:\",filename)\n",
    "    count = 0\n",
    "    for v in scores:\n",
    "        #print(\"Score:\", v)\n",
    "        if count ==0:\n",
    "            match = int(v)\n",
    "        elif count == 1:\n",
    "            mismatch = int(v)\n",
    "        elif count == 2:\n",
    "            gapex = int(v)\n",
    "        elif count == 3:\n",
    "            gapop = int(v)\n",
    "        count += 1\n",
    "        \n",
    "    print(\"MATCH: \", match, \" MISMATCH: \", mismatch, \" GAPEX: \", gapex, \" GAPOP: \", gapop)\n",
    "\n",
    "    fin = str(match)+ \";\"+str(mismatch)+\";\"+ str(gapex)+\";\"+ str(gapop)\n",
    "    print(fin)\n",
    "    print()\n",
    "    return fin\n",
    "'''\n",
    "#GLOBAL ALIGNMENT WITH AFFINE GAPS\n",
    "#This function uses a dynamic programming algorithm to calculate the optimal\n",
    "#alignment score between two strings using gap, mismatch, and match scores\n",
    "#Parameters: s1 and s2 are two strings (do not have to be the same length)\n",
    "#            gap, mismatch, and match are all integers, gap and mismatch are\n",
    "#            typically negative values, match is typically a positive value\n",
    "#Returns: the optimal global alignment score, as we as a 2-D list of hints\n",
    "#         to help determine the actual alignment\n",
    "'''\n",
    "def globalAF(s1, s2, gapex, mismatch, match, gapop):\n",
    "    m = len(s1)\n",
    "    n = len(s2)\n",
    "\n",
    "    #Initialize 2 2-d lists of size (m+1)rows * (n+1)columns\n",
    "    gapy = [[0 for i in range(n+1)] for j in range(m+1)]  #gapy\n",
    "    matches = [[0 for i in range(n+1)] for j in range(m+1)]   #matches\n",
    "    gapx = [[0 for i in range(n+1)] for j in range(m+1)]   #gapx\n",
    "    \n",
    "    matrix = [[0 for i in range(n+1)] for j in range(m+1)]   #final/optimal matrix\n",
    "    \n",
    "    hints = [[0 for i in range(n+1)] for j in range(m+1)]   #backtracking hints\n",
    "    \n",
    "    gapex = int(gapex)\n",
    "    gapop = int(gapop)\n",
    "    \n",
    "    #Below is the pseudocode for computing the alignment score table (matrix)\n",
    "    #Once you get this working, add in code to keep track of the hints\n",
    "    #You'll see in get_alignment, that it is expecting you to present hints as:\n",
    "    # \\\\ (diagonal), | (from above), - (from left)\n",
    "    \n",
    "    #initalization of matrices        \n",
    "    for i in range(0, m+1): #first row\n",
    "        matrix[i][0] = gapop + (gapex * i)\n",
    "        gapy[i][0] = gapop + (gapex * i)\n",
    "        \n",
    "        #invalid\n",
    "        gapx[i][0] = -999999999999\n",
    "        matches[i][0] = -9999999999\n",
    "        \n",
    "        hints[i][0] = \"|\"\n",
    "\n",
    "    for j in range(0, n+1):\n",
    "        matrix[0][j] = gapop + (gapex * j)\n",
    "        gapx[0][j] = gapop + (gapex * j)\n",
    "        \n",
    "        #invalid\n",
    "        gapy[0][j] = -999999999999\n",
    "        matches[0][j] = -9999999999\n",
    "        \n",
    "        hints[0][j] = \"-\"\n",
    "\n",
    "    #initalize the [0][0] corner\n",
    "    matrix[0][0] = 0  \n",
    "    matches[0][0] = 0\n",
    "    gapy[0][0] = gapop \n",
    "    gapx[0][0] = gapop\n",
    "    \n",
    "    #new gap (gapop + gapex) if coming from match/mismatch\n",
    "    #other wise extend (gapex)\n",
    "    for i in range(1, m+1):\n",
    "        for j in range(1, n+1):\n",
    "            #x and y match/mismatch\n",
    "            if s1[i-1] == s2[j-1]:\n",
    "                score1_1 = matches[i-1][j-1] #prev match\n",
    "                score1_2 = gapy[i-1][j-1] #prev gap in y direction\n",
    "                score1_3 = gapx[i-1][j-1] #prev gap in x direction\n",
    "                score1 = max(matches[i-1][j-1], gapy[i-1][j-1], gapx[i-1][j-1]) + match\n",
    "                matches[i][j] = score1\n",
    "                #print(\"MAX: \", max(matches[i-1][j-1], gapy[i-1][j-1], gapx[i-1][j-1]))\n",
    "            else:\n",
    "                score1 = max(matches[i-1][j-1], gapy[i-1][j-1], gapx[i-1][j-1]) + mismatch\n",
    "                matches[i][j] = score1\n",
    "            \n",
    "            #gap in the x direction\n",
    "            score2_1 = gapex + gapop + matches[i][j-1] \n",
    "            score2_2 = gapex + gapx[i][j-1]\n",
    "            score2_3 = gapex + gapop + gapy[i][j-1]\n",
    "            \n",
    "            score2 = max(score2_1, score2_2, score2_3)\n",
    "            gapx[i][j] = score2\n",
    "            \n",
    "            #gap in the y direction\n",
    "            score3_1 = gapex + gapop + matches[i-1][j]\n",
    "            score3_2 = gapex + gapop + gapx[i-1][j]\n",
    "            score3_3 = gapex + gapy[i-1][j]\n",
    "            \n",
    "            score3 = max(score3_1, score3_2, score3_3)\n",
    "            gapy[i][j] = score3\n",
    "            \n",
    "            \n",
    "            #max score overall\n",
    "            maxScore = max(score1, score2, score3)\n",
    "            \n",
    "            #add to optimal matrix and add direction to hints\n",
    "            matrix[i][j] = maxScore\n",
    "            if maxScore == score1:\n",
    "                hints[i][j] = \"\\\\\"\n",
    "            elif maxScore == score2:\n",
    "                hints[i][j] = \"-\"\n",
    "            else:\n",
    "                hints[i][j] = \"|\"\n",
    "            \n",
    "#     print(\"matrix /S\")\n",
    "#     for row in matrix:\n",
    "#         print(row)\n",
    "#     print(\"gap y /F\")\n",
    "#     for row in gapy:\n",
    "#         print(row)\n",
    "#     print(\"gap x /E\")\n",
    "#     for row in gapx:\n",
    "#         print(row)\n",
    "#     print(\"matches / G\")\n",
    "#     for row in matches:\n",
    "#         print(row)\n",
    "#     print(\"hints\")\n",
    "#     for row in hints:\n",
    "#         print(row)\n",
    "    return matrix[m][n], hints\n",
    "\n",
    "#This function takes in 2 strings and a 2-D list of backtracking hints\n",
    "#and returns the global alignment of the strings\n",
    "#Parameters: A and B are two strings (do not have to be the same length)\n",
    "#            hints is a 2-D lists containing \\\\, -, or | symbols\n",
    "#Returns: two strings describing the global alignment between A and B\n",
    "\n",
    "#requires 2 sequences, prints alignment and scores\n",
    "def globalalignAF(seq1, seq2):\n",
    "    s1 = seq1\n",
    "    s2 = seq2\n",
    "    option = 0\n",
    "    opvalid = False\n",
    "    while not opvalid:\n",
    "        option = input(\"For penalties, Filename (1) or by hand (2)? \")\n",
    "        if option == \"1\":\n",
    "            filename = input(\"Filename: \")\n",
    "            scores = str(scoringfile(filename))\n",
    "            match, mismatch, gapex, gapop = scores.split(\";\")\n",
    "            gap_score = int(gapex)\n",
    "            mismatch_score = int(mismatch)\n",
    "            match_score = int(match)\n",
    "            gapop = int(gapop)\n",
    "            opvalid = True\n",
    "        elif option == \"2\":\n",
    "            mismatch_score = int(input(\"Mismatch: \"))\n",
    "            match_score = int(input(\"Match: \"))\n",
    "            gapop = int(input(\"Gap open: \"))\n",
    "            gap_score = int(input(\"Gap extend: \"))\n",
    "            opvalid = True\n",
    "        else:\n",
    "            option = input(\"Try again. Filename (1) or by hand (2)? \")\n",
    "            opvalid = False\n",
    "    \n",
    "    score, hints = globalAF(s1, s2, gap_score, mismatch_score, match_score, gapop)\n",
    "    print(\"The global alignment with afine gaps score is\", score)\n",
    "    \n",
    "    align1, align2 = get_alignment(s1, s2, hints)\n",
    "    print(align1)\n",
    "    print(align2)\n",
    "   "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "33a379b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "'''LOCAL ALIGNMENT'''\n",
    "\n",
    "'''\n",
    "#This function uses a dynamic programming algorithm to calculate the optimal\n",
    "#alignment score between two strings using gap, mismatch, and match scores\n",
    "#Parameters: s1 and s2 are two strings (do not have to be the same length)\n",
    "#            gap, mismatch, and match are all integers, gap and mismatch are\n",
    "#            typically negative values, match is typically a positive value\n",
    "#Returns: the optimal local alignment score, as we as a 2-D list of hints\n",
    "#         to help determine the actual alignment\n",
    "'''\n",
    "def SmithWaterman(s1, s2, gap, mismatch, match, gapop):\n",
    "    m = len(s1)\n",
    "    n = len(s2)\n",
    "\n",
    "    # Initialize 2 2-d lists of size (m+1)rows * (n+1)columns\n",
    "    matrix = [[0 for i in range(n + 1)] for j in range(m + 1)]  # scores\n",
    "    hints = [[0 for i in range(n + 1)] for j in range(m + 1)]  # backtracking hints\n",
    "\n",
    "    # \\\\ (diagonal), | (from above), - (from left)\n",
    "\n",
    "    matrix[0][0] = 0\n",
    "    for i in range(1, m+1):\n",
    "        matrix[i][0] = 0\n",
    "        hints[i][0] = \"|\"\n",
    "\n",
    "    for j in range(1, n+1):\n",
    "        matrix[0][j] = 0\n",
    "        hints[0][j] = \"-\"\n",
    "\n",
    "    # checks the recurrences\n",
    "    for i in range(1, m+1):\n",
    "        for j in range(1, n+1):\n",
    "            if s1[i-1] == s2[j-1]:\n",
    "                score1 = matrix[i-1][j-1] + match\n",
    "            else:\n",
    "                score1 = matrix[i-1][j-1] + mismatch\n",
    "\n",
    "            # all the negative values become 0\n",
    "            if score1 < 0:\n",
    "                score1 = 0\n",
    "\n",
    "            score2 = matrix[i][j-1] + gap\n",
    "            if score2 < 0:\n",
    "                score2 = 0\n",
    "\n",
    "            score3 = matrix[i-1][j] + gap\n",
    "            if score3 < 0:\n",
    "                score3 = 0\n",
    "\n",
    "            maxScore = max(score1, score2, score3)\n",
    "            matrix[i][j] = maxScore\n",
    "\n",
    "            if maxScore == score1:\n",
    "                hints[i][j] = \"\\\\\"\n",
    "            elif maxScore == score2:\n",
    "                hints[i][j] = \"-\"\n",
    "            elif maxScore == score3:\n",
    "                hints[i][j] = \"|\"\n",
    "            else: # meaning that the score is 0\n",
    "                hints[i][j] = \".\"\n",
    "                \n",
    "    return matrix[m][n], hints\n",
    "\n",
    "\n",
    "def get_alignment(A, B, hints):\n",
    "    i = len(A)\n",
    "    j = len(B)\n",
    "    align1 = ''\n",
    "    align2 = ''\n",
    "\n",
    "    while i > 0 or j > 0:\n",
    "        if hints[i][j] == '\\\\':\n",
    "            align1 = A[i - 1] + align1\n",
    "            align2 = B[j - 1] + align2\n",
    "            i -= 1\n",
    "            j -= 1\n",
    "        elif hints[i][j] == '|':\n",
    "            align1 = A[i - 1] + align1\n",
    "            align2 = \"-\" + align2\n",
    "            i -= 1\n",
    "        elif hints[i][j] == '.':\n",
    "            break\n",
    "        else:\n",
    "            align1 = \"-\" + align1\n",
    "            align2 = B[j - 1] + align2\n",
    "            j -= 1\n",
    "\n",
    "    print(i, j)\n",
    "    return align1, align2\n",
    "\n",
    "def localalign(seq1, seq2):\n",
    "    s1 = seq1\n",
    "    s2 = seq2\n",
    "    option = 0\n",
    "    while option != \"1\" and option != \"2\":\n",
    "        option = input(\"For penalties, Filename (1) or by hand (2)? \")\n",
    "        if option == \"1\":\n",
    "            filename = input(\"Filename: \")\n",
    "            scores = str(scoringfile(filename))\n",
    "            match, mismatch, gapex, gapop = scores.split(\";\")\n",
    "            gap_score = int(gapex)\n",
    "            mismatch_score = int(mismatch)\n",
    "            match_score = int(match)\n",
    "            gapop = int(gapop)\n",
    "        elif option == \"2\":\n",
    "            mismatch_score = int(input(\"Mismatch: \"))\n",
    "            match_score = int(input(\"Match: \"))\n",
    "            gapop = int(input(\"Gap open: \"))\n",
    "            gap_score = int(input(\"Gap extend: \"))\n",
    "        else:\n",
    "            option = input(\"Try again. Filename (1) or by hand (2)? \")\n",
    "    \n",
    "    score, hints = SmithWaterman(s1, s2, gap_score, mismatch_score, match_score, gapop)\n",
    "    print(\"The local alignment is\", score)\n",
    "    \n",
    "    align1, align2 = get_alignment(s1, s2, hints)\n",
    "    print(align1)\n",
    "    print(align2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "06437282",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CAT/HUMAN\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The global alignment with afine gaps score is -26\n",
      "0 0\n",
      "TTAATGAAATGATACACGTCTTGCCCCCAAACAGTTGTTTTATTACACGATGTCTGAGACTTGGTGCCTTCACGGGT-GCTGATTCCCTAACGTC--CCTGTTG-GTGGATGGGATGATCAAGTTCCTACGACC-GGGACAGCTGGTCAGGGGTCCCCTCCTTGCTGGCCCGGCCGAATC-CTGTCTTCCCTCTCGCCGAACCTC-ACTTCCTAAAAGGAAGTCTGACTCTGTCCTACGCCCGCGAGGGATCCGATCCCTG-GTGTAGGATCTGTATCTATCAAGGTCTGTTCC-CTGCCCCCGACTCCTCAGGATCAATTTCTTTTGCACTCTTGGAGGCACATCTGCGTATTTCCCCACACCCTT--CC-CC--CT--C--------TCCCCTTTTTATATC---C-CCT-TTTT---AT---A------T-----C-G-GTTTC------GT-\n",
      "--CA-GG-CTG-GAGTGGAG-TGGCGTG-ATC--TTG---GCTT--AC--TG-C--AGCCTT--TGCCTCCCCGGCTCGAGCAGT-CCT-GCCTCAGCCTCCGGAGTAGCTGGGACCA-CAGGTTCATGCCACCATGGCCAGC----CAACTTTTGCATGTTTTGTAGAGATGGGGTCTCACAGTGTTGCC-CAGGCTG-GTCTCAAACTCCT----GG--G-CTCAGGCGATCC-A--CCTGTCTCAG--CC--TCCCAGAGTGCTGG----G-AT-TA-CAA-TTGTGAGCCACCACGTCCAGCT-GGAAGGGTCAACATCTTTTACATTC-TGCAAGCACATCTGCAT-TTTCACCCCACCCTTCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCACCTGTGTGTCTGAGGGGTG\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The local alignment is 54\n",
      "0 0\n",
      "TTAATGAAATGATACACGTCTTGCCCCCAAACAGTTGTTTTATTACACGATGTCTGAGACTTGGTGCCTTCACGGGTGCTGATTCCCTAACGTCCCTGTTG-GTGGA-TGG-GATGATCAAGTTCCTAC-G-A-CC-GGGAC-AGCTGG-TC-AGGGGTCC--CCTC--CTTGCTG-G--CCCGGCCGA--ATCCTG-TC-TTCC-CTC-TCGCC-G--AAC--CT-CA--CTTCCTA-A-AAGGAAGTCTGACTCTGTCCTACGCCC-GCGAGGGATC-CGATC-CCTGGTG-T-A-G-GAT----CTGTATCTA---T-CAAG-GT-CT--G-TT-C---CCT--GCCCCCGAC-T-C--CT--CAGGATCAATTTCTTTTGCACTCTTGGAGGCACATCTGCGTATTTC-CCCACACCCTT-CCC-CC-TCT--C---------CCC-TTTTTATATC---C-CCT-TTTT---AT---A---T-C-G------GT-T-TC------GT-\n",
      "-------------------------------------------------------------------------------------------CAGGCTG--GAGTGGAGTGGCG-TGATCTTG-GCTTACTGCAGCCTTTGCCTCCCCGGCTCGAGCAGTCCTGCCTCAGCCTCCGGAGTAGCTGG--GACCA-CAGGTTCATGCCAC-CATGGCCAGCCAACTTTTGCATGTTTTGTAGAGATGG-GGTCTCACAGTGT--T--GCCCAG-GCTGG-TCTCAAACTCCTGG-GCTCAGGCGATCCACCTGTCTC-AGCCTCCCAGAGTGCTGGGATTACAATTGTGAGCCACC-ACGTCCAGCTGGAAGGGTCAACATCTTTTACATTC-TGCAAGCACATCTGCAT-TTTCACCC-CACCCTTCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCACCTGTGTGTCTGAGGGGTG\n",
      "CAT/RAT\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The global alignment with afine gaps score is -74\n",
      "0 0\n",
      "TTA--ATG-AAAT-G-ATACA-CGTCTTGCCCCCAAACAGT-T----GTT-TT-AT-TACAC-GATGTCTGAGACTTGGTGC-CTTCACGGGTGCTGATTCCCTAACGTCCCTGTTGGTGGAT-GGGA-TGATCAAGTTCCTACGACCGGGACAGCTGGTCAGGGGTCCCCT-CCTTGCTGGCCCGGCCGAATCCTGTCTTCCCTCTCGCCGAACCTCACTTCCTAAAAGGAAGTCTGACTCTGTCCTACGCCCGCGAGGGA-TCCGATCCCTGGTGTAGG---AT--CTGTATCTATCAAGGTCTGTTCCCTGCCCCCGACTCCTCAGGATCAATTTCTTTTGCACTCTTGGAGGCACA-TCTGCGTATTTCCCCACACCCTTCCCCCTCTC-CCCTTTTTATATCCC-CTTTT---TA-TATCGGTTTCGT\n",
      "GTACGAGGCAACTGGAAAAAATAGCAATG-TATTAAATAGTGTACGAGTTCTTAATCTCCACAAATTTCCGTTGC-T-GTGCAATTGAAGGCTG-TGA-AAACAATCTTTACAGATAGT-GATAAGGACTGTACAGGAAATTA--AAC---A-AGGTGGTGTGATATAAAGTATCTCGGAGAACAGG-TTAGT-TAGATTGAGATAT-TAGGAAAAGC-CTTTCTAAAGTGACATTTTA-GCTG--ATGAGGGGGAAAACAATTTAGCTGCC-CGTGTTGGTTCATCCCTGTA-CTTGGAAGGCCT-AAAGCAGG--AAGACTGC-C-GCA--AATCGC-AGGCCAGCCCTGGCTACAAAGACT-C-TGTTT-CAAAAAGCCAAAAAGATGACTATGATTATCTAGCTGGATAGGAAAGAGCAGCGCAGTGGC\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The local alignment is 0\n",
      "0 0\n",
      "-------TTAA-T-GAAATGATA-C-ACGTCTTGCCCCCAAACAGT-T----GTT-TT-AT-TACAC-GATGTCTG-AGACTTGGTGC-CTTCACGGGTGCTGATTCCCTAACGTCCCTGTTGGTGGATGGGATGATCAAGTTCCTACG-ACCGGGACAGCTGGTC-AGG-GGTCCCCTCCT--TGCTGGCCCGGCCGAATC---CT-G-T--CTT-CCCT-CTCGCCGAACCTCACTTCCTAAAAG-GA-AGTCTGACTCTG-TCCTACGCCCGCGAGGGATCCGA--TCCCTG---GTGTAGG---AT--CTGTATCTAT-CAAGGTCTGTTCCCTGCCCCCGACTCCTCAGGATCAATTTCTTTTG-CA-CTCTTGGAGGCACA-TCTGCGTATTTCCCCACACCCTTCCCCCT--CTCCCCTT-TTTA--T--ATCCCCTTTTTATATCGGTTTCGT\n",
      "GTACGAGGCAACTGGAAAAAATAGCAATGTATT------AAATAGTGTACGAGTTCTTAATCTCCACAAATTTCCGTTG-C-T-GTGCAATTGAAGGCTG-TGA-AAAC-AA--T--CT-TT-ACAGAT-AG-TGAT-AAG-GACT--GTA-CAGGA-AATTAAACAAGGTGGTGTGAT-ATAAAG-TATCTCGG-AGAA-CAGGTTAGTTAGATTGAGATATTAG--GAAAAGC-CTTTCT-AAAGTGACA-TTTTA-GCTGAT--GA-G---G-G-GGAAAACAATTTAGCTGCCCGTGTTGGTTCATCCCTGTA-CT-TGGAAGGCCT-AAAGCAG--GAAGACTGC-C--G--CAA-ATCGCAGGCCAGC-CCTGGCTACAAAGACT-C-TGTTT-CAAAAAGCCAAAAAGATGACTATGATTATCTAGCTGGAT-AGGAAAGAGCAGCGCAGTGGC\n",
      "RAT/HUMAN\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The global alignment with afine gaps score is -61\n",
      "0 0\n",
      "GTACGAGGCAACTGGA-AAAAATAGCAATGTATTAAATAGTGTAC-G-AGTTCTTAATCTCCACAAAT--TTCCGT--TG-CTGTGCAATTGAAG--GCTGTGAAAACAATCTTTACAGATAGTGATAAGGACTGTACAG-GAAATTAAACAAGGTGGTGT-GATATAAAGTATCTC---GGAGAACAGGTTAGTTAGATTGAGATATTAGGAAAAGCCTTTCTA-AAGTGACATTTTAGCTGA-TGAGGGGGAAAACAATT-T-AGCTGCC-CGTGTTGGTTCATCCCTGTACTTGGAAGGCCTAAAGCAGGAAGACTGCCGCAAATCGCAGGCCAGCCCTGGCTACAAAGACTCTGTTTCAAAAAGCCA--AAAAGAT-GA-CTATGA--TT---ATCTAGC-TGGATAGGA-AAGAGCAGCGCAGTGG-C\n",
      "---C-AGG---CTGGAGTGGAGTGGC---GTGAT-CTTGGCTTACTGCAG-CCTTTGCCTCCCCGGCTCGAGCAGTCCTGCCTCAGCCTCCGGAGTAGCTGGGACCAC-A-GGTT-C----A-TG-CCACCA-TGGCCAGCCAACTTTTGC-ATGTTTTGTAGAGATGGGGTCTCACAGTGTTGCCCAGGCT-GGT--CTCAAACTCCTGGGCTCAGGCGATCCACCTGTCTCAGCCTCCCAGAGTG-CTGGGATTACAATTGTGAGCCACCACGTCCAGCTGGAAGGGTCAACATCTTTTACATTCTGCAAGCACA-T-CTGC-ATTTTCACCCCA-CCCT-TCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATTTTACAATAAAACTTTGCTGCCACCTGTGTGTCTGAGGGGTG\n"
     ]
    },
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "For penalties, Filename (1) or by hand (2)?  1\n",
      "Filename:  BRCAscoreA.txt\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Printing contents of following file: BRCAscoreA.txt\n",
      "#match\tmismatch\tgap-extend\tgap-open\n",
      "1\t-1\t-1\t-1\n",
      "\n",
      "\n",
      "\n",
      "\n",
      "Scores contained in: BRCAscoreA.txt\n",
      "MATCH:  1  MISMATCH:  -1  GAPEX:  -1  GAPOP:  -1\n",
      "1;-1;-1;-1\n",
      "\n",
      "The local alignment is 13\n",
      "0 0\n",
      "GTACGAGGCAACTGGAAAAAATAGCAATGTATTAAATAGTGTACGAGTTCTT--AAT-CTCCA-CAAATTT-CCGTTGCTGTGCAATTGA--AGGCTGTGAAAACAATCTTTACAGA-TAG-T--GA-TA-AGG-AC-TG-TA-CA-GG--A---AA-TTAAACAAGGTGGTGT-GATATAAAGTATCTCGGAGAACAG-GTTAG-TTAGATTGAGATATTAGGAAAAGCCT-TTCTAAAGTGA--CA--T-T-TTAG-CT--GATGAG-G--GGGAAAACAATT-T-AGCTGCC-CGT---G-T---TGGTTCATC-CCTGTACTTGGA-AGGCCTAAAGCAGGAAG-AC-TGCCGCAAATCGCAGGCCAGCCCTGGCTACAAAGACTCTGTTTCAAAAAGCCA--AAAAGAT-GA-CTATGATTATCTAGCTGGATAGGAAA----G-AG-CA-----GCG-C--AGTGG-C\n",
      "------------------------CAGGCTGGAGTGGAGTG-GCGTGATCTTGGCTTACTGCAGC--CTTTGCC-TCCCCG-GC--TCGAGCAGTC-CTG-CCTC-AGC-CTCCGGAGTAGCTGGGACCACAGGTTCATGCCACCATGGCCAGCCAACTTTTGC-ATGTTTTGTAGAGAT--GGGGTCTC-----ACAGTGTT-GCCCAGGCTG-G-T-CT--CAAACTCCTGGGCTCAGGCGATCCACCTGTCTCAGCCTCCCA-GAGTGCTGGGATTACAATTGTGAGCCACCACGTCCAGCTGGAAGGGTCAACATCT-T--TT--ACA-TTCT---GC---AAGCACAT-CTGC-ATTTTCACCCCA-CCCT-TCCCCTCCTTCTCCCTTTTTATATCCCATTTTTATATCGATCTCTTATT-T-TA-C--AATA--AAACTTTGCTGCCACCTGTGTGTCTGAGGGGTG\n"
     ]
    }
   ],
   "source": [
    "'''\n",
    "HUMAN/CAT/RAT genome comparisons\n",
    "'''\n",
    "\n",
    "cat = read_single_fasta(\"TP53-Cat.fasta\")\n",
    "human = read_single_fasta(\"TP53-Human.fasta\")\n",
    "rat = read_single_fasta(\"TP53-Rat.fasta\")\n",
    "\n",
    "print(\"CAT/HUMAN\")\n",
    "globalalignAF(cat, human)\n",
    "localalign(cat, human)\n",
    "print(\"CAT/RAT\")\n",
    "globalalignAF(cat, rat)\n",
    "localalign(cat, rat)\n",
    "print(\"RAT/HUMAN\")\n",
    "globalalignAF(rat, human)\n",
    "localalign(rat, human)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "50001a9a",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = read_single_fasta(\"BRCA1-exons.txt\")\n",
    "partial = read_single_fasta(\"BRCA1-partial.txt\")\n",
    "\n",
    "globalalignAF(exons, partial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "627c0faf",
   "metadata": {},
   "outputs": [],
   "source": [
    "exons = read_single_fasta(\"BRCA1-exons.txt\")\n",
    "partial = read_single_fasta(\"BRCA1-partial.txt\")\n",
    "\n",
    "localalign(exons, partial)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b6f4c8ce",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.15"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
